File: backend/src/main/java/com/suplr/backend/service/WhatsAppSenderService.java
package com.suplr.backend.service;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.reactive.function.client.WebClientResponseException;

import java.time.Duration;
import java.time.OffsetDateTime;
import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Slf4j
@Service
public class WhatsAppSenderService {

    private WebClient client(String bspEndpoint, String bspApiKey) {
        return WebClient.builder()
                .baseUrl(bspEndpoint.replaceAll("/+$", ""))
                .defaultHeader("Content-Type", "application/json")
                .defaultHeader("Authorization", "Bearer " + bspApiKey)
                .build();
    }

    public void sendMessage(String bspEndpoint, String bspApiKey, String to, String text) {
        sendMessage(bspEndpoint, bspApiKey, to, text, null);
    }

    public void sendMessage(String bspEndpoint, String bspApiKey,
                            String to, String text, String mediaUrl) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("to", to);
        payload.put("message", text);
        if (mediaUrl != null) payload.put("mediaUrl", mediaUrl);

        try {
            client(bspEndpoint, bspApiKey)
                    .post().uri("/send")
                    .bodyValue(payload)
                    .retrieve()
                    .toBodilessEntity()
                    .timeout(Duration.ofSeconds(10))
                    .block();
        } catch (WebClientResponseException e) {
            log.error("BSP send failed to {}: {} {}", to, e.getStatusCode(), e.getResponseBodyAsString());
            throw new RuntimeException("BSP send failed: " + e.getMessage(), e);
        } catch (Exception e) {
            log.error("BSP send error to {}: {}", to, e.getMessage());
            throw new RuntimeException("BSP send error: " + e.getMessage(), e);
        }
    }

    public void sendDocument(String bspEndpoint, String bspApiKey,
                             String to, byte[] pdfBytes, String filename) {
        Map<String, Object> payload = new HashMap<>();
        payload.put("to", to);
        payload.put("filename", filename);
        payload.put("base64", Base64.getEncoder().encodeToString(pdfBytes));

        try {
            client(bspEndpoint, bspApiKey)
                    .post().uri("/send-document")
                    .bodyValue(payload)
                    .retrieve()
                    .toBodilessEntity()
                    .timeout(Duration.ofSeconds(30))
                    .block();
        } catch (Exception e) {
            log.error("BSP send-document failed to {}: {}", to, e.getMessage());
            throw new RuntimeException("BSP send-document failed: " + e.getMessage(), e);
        }
    }

    public String enqueueBroadcast(String bspEndpoint, String bspApiKey,
                                   List<String> numbers, String message,
                                   OffsetDateTime scheduledAt, String mediaUrl) {
        long delayMs = 0;
        if (scheduledAt != null) {
            long seconds = Duration.between(OffsetDateTime.now(), scheduledAt).toSeconds();
            delayMs = Math.max(0, seconds * 1000);
        }

        Map<String, Object> payload = new HashMap<>();
        payload.put("numbers", numbers);
        payload.put("message", message);
        payload.put("delayMs", delayMs);
        if (mediaUrl != null) payload.put("mediaUrl", mediaUrl);

        try {
            Map<?, ?> response = client(bspEndpoint, bspApiKey)
                    .post().uri("/queue/broadcast")
                    .bodyValue(payload)
                    .retrieve()
                    .bodyToMono(Map.class)
                    .timeout(Duration.ofSeconds(10))
                    .block();
            return (response != null && response.get("jobId") != null)
                    ? response.get("jobId").toString()
                    : "";
        } catch (Exception e) {
            log.error("BSP enqueue-broadcast failed: {}", e.getMessage());
            throw new RuntimeException("BSP enqueue-broadcast failed: " + e.getMessage(), e);
        }
    }

    public String enqueueReminder(String bspEndpoint, String bspApiKey,
                                  String number, String message,
                                  OffsetDateTime fireAt, String jobId) {
        long seconds = Duration.between(OffsetDateTime.now(), fireAt).toSeconds();
        long delayMs = Math.max(0, seconds * 1000);

        Map<String, Object> payload = new HashMap<>();
        payload.put("number", number);
        payload.put("message", message);
        payload.put("delayMs", delayMs);
        if (jobId != null) payload.put("jobId", jobId);

        try {
            Map<?, ?> response = client(bspEndpoint, bspApiKey)
                    .post().uri("/queue/reminder")
                    .bodyValue(payload)
                    .retrieve()
                    .bodyToMono(Map.class)
                    .timeout(Duration.ofSeconds(10))
                    .block();
            return (response != null && response.get("jobId") != null)
                    ? response.get("jobId").toString()
                    : "";
        } catch (Exception e) {
            log.error("BSP enqueue-reminder failed: {}", e.getMessage());
            throw new RuntimeException("BSP enqueue-reminder failed: " + e.getMessage(), e);
        }
    }
}

File: backend/src/whatsapp/whatsapp-sender.service.ts
import { Injectable, Logger } from '@nestjs/common';
import axios from 'axios';

@Injectable()
export class WhatsAppSenderService {
  private readonly logger = new Logger(WhatsAppSenderService.name);

  private headers(bspApiKey: string) {
    return { 'Content-Type': 'application/json', Authorization: `Bearer ${bspApiKey}` };
  }

  async sendMessage(bspEndpoint: string, bspApiKey: string, to: string, text: string, mediaUrl?: string): Promise<void> {
    const payload: Record<string, unknown> = { to, message: text };
    if (mediaUrl) payload.mediaUrl = mediaUrl;
    try {
      await axios.post(`${bspEndpoint.replace(/\/+$/, '')}/send`, payload, { headers: this.headers(bspApiKey), timeout: 10000 });
    } catch (e: any) {
      this.logger.error(`BSP send failed to ${to}: ${e.message}`);
      throw new Error(`BSP send failed: ${e.message}`);
    }
  }

  async sendDocument(bspEndpoint: string, bspApiKey: string, to: string, pdfBytes: Buffer, filename: string): Promise<void> {
    const payload = { to, filename, base64: pdfBytes.toString('base64') };
    try {
      await axios.post(`${bspEndpoint.replace(/\/+$/, '')}/send-document`, payload, { headers: this.headers(bspApiKey), timeout: 30000 });
    } catch (e: any) {
      this.logger.error(`BSP send-document failed to ${to}: ${e.message}`);
      throw new Error(`BSP send-document failed: ${e.message}`);
    }
  }

  async enqueueBroadcast(bspEndpoint: string, bspApiKey: string, numbers: string[], message: string, scheduledAt?: Date | null, mediaUrl?: string): Promise<string> {
    let delayMs = 0;
    if (scheduledAt) {
      delayMs = Math.max(0, (new Date(scheduledAt).getTime() - Date.now()));
    }
    const payload: Record<string, unknown> = { numbers, message, delayMs };
    if (mediaUrl) payload.mediaUrl = mediaUrl;
    try {
      const res = await axios.post(`${bspEndpoint.replace(/\/+$/, '')}/queue/broadcast`, payload, { headers: this.headers(bspApiKey), timeout: 10000 });
      return (res.data as any)?.jobId ?? '';
    } catch (e: any) {
      this.logger.error(`BSP enqueue-broadcast failed: ${e.message}`);
      throw new Error(`BSP enqueue-broadcast failed: ${e.message}`);
    }
  }
}
